python/pip_install/extract_wheels/lib/namespace_pkgs.py

```diff
@@ -19,26 +20,30 @@ def implicit_namespace_packages(
     Returns:
         The set of directories found under root to be packages using the native namespace method.
     """
-    namespace_pkg_dirs = set()
-    for dirpath, dirnames, filenames in os.walk(directory, topdown=True):
-        # We are only interested in dirs with no __init__.py file
+    namespace_pkg_dirs: Set[str] = set()
+    standard_pkg_dirs: Set[str] = set()
+    # Traverse bottom-up because a directory can be a namespace pkg because its child contains module files.
+    for dirpath, dirnames, filenames in os.walk(directory, topdown=False):
         if "__init__.py" in filenames:
-            dirnames[:] = []  # Remove dirnames from search
+            standard_pkg_dirs.add(dirpath)
             continue
+        elif ignored_dirnames:
+            is_ignored_dir = dirpath in ignored_dirnames
+            child_of_ignored_dir = any(d in pathlib.Path(dirpath).parents for d in ignored_dirnames)
+            if is_ignored_dir or child_of_ignored_dir:
+                continue
 
-        for ignored_dir in ignored_dirnames or []:
-            if ignored_dir in dirnames:
-                dirnames.remove(ignored_dir)
-
-        non_empty_directory = dirnames or filenames
+        dir_includes_py_modules = _includes_python_modules(filenames)
+        parent_of_namespace_pkg = any(str(pathlib.Path(dirpath, d)) in namespace_pkg_dirs for d in dirnames)
+        parent_of_standard_pkg = any(str(pathlib.Path(dirpath, d)) in standard_pkg_dirs for d in dirnames)
+        parent_of_pkg = parent_of_namespace_pkg or parent_of_standard_pkg
```

Comment:
nit: if this is the only use of this variable it seems cleaner to put it in the conditional where it is used.